tensorflow/compiler/jit/xla_compilation_cache.cc

```diff
@@ -181,33 +181,37 @@ Status XlaCompilationCache::BuildExecutable(
 
 Status XlaCompilationCache::Compile(
     const XlaCompiler::Options& options, const NameAttrList& function,
-    absl::Span<const XlaCompiler::Argument> args,
+    std::vector<XlaCompiler::Argument>& args,
     const XlaCompiler::CompileOptions& compile_options,
     CompileMode compile_mode,
     const XlaCompiler::CompilationResult** out_compilation_result,
     xla::LocalExecutable** out_executable) {
-  absl::optional<int64> compile_threshold;
-  if (compile_mode == CompileMode::kLazy) {
-    compile_threshold = kDefaultCompilationThreshold;
-  }
-  auto compile_fn = [&](XlaCompiler* compiler,
+  // compile_fn can be called asynchronously. Make sure all required arguments
+  // are passed by value.
+  auto compile_fn = [&, compile_options, function](
+                        XlaCompiler* compiler,
+                        const std::vector<XlaCompiler::Argument>& args,
```

Comment:
@sanjoy, I'm not seeing how you want me to use the operator().
please add some pseudo code to show the intent

Reply:
I meant writing this as:

```
struct CompileFunctor {
  CompileOptions compile_options;
  // All the other state that is needed
  
  Status operator()(<args>) { ... }
};
```

And create an instance of `CompileFunctor` and pass that to `CompileImpl`.  The difference is that the state captured is now explicit.  In the current version it will be easy for a later change to introduce the use of a local variable in the body of the lambda.

Reply:
Passing additional to CompileImpl requires many additional functions tot changes as well. Capture by value is cleaner.
Added a POD-type to explicitly capture the required variables, added additional comments for clarity, and most importantly, removed capture-default to prevent  additional variables to be introduced in the body of the lambda